locker.go

```diff
@@ -0,0 +1,180 @@
+package goose
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/sethvargo/go-retry"
+)
+
+// Locker defines the methods to lock and unlock the database.
+//
+// Locking is an experimental feature and the underlying implementation may change in the future.
+//
+// The only database that currently supports locking is Postgres. Other databases will return
+// ErrLockNotImplemented.
+type Locker interface {
+	// LockSession and UnlockSession are used to lock the database for the duration of a session.
+	//
+	// The session is defined as the duration of a single connection and both methods must be called
+	// on the same connection.
+	LockSession(ctx context.Context, conn *sql.Conn) error
+	UnlockSession(ctx context.Context, conn *sql.Conn) error
+
+	// LockTransaction is used to lock the database for the duration of a transaction.
+	// LockTransaction(ctx context.Context, tx *sql.Tx) error
+}
+
+var (
+	// ErrLockNotImplemented is returned when the database does not support locking.
+	ErrLockNotImplemented = errors.New("lock not implemented")
+)
+
+type PostgresLocker struct {
+	retryLock   retry.Backoff
+	retryUnlock retry.Backoff
+	opts        PostgresLockerOptions
+}
+
+type PostgresLockerOptions struct {
+	LockID int64
+}
+
+func NewPostgresLocker(opts PostgresLockerOptions) *PostgresLocker {
```

Comment:
Earlier, this locker was encapsulated (by putting it in an `internal` package). 

But it can't be encapsulated if we want to allow injecting this provider (and configuration).